slf4j-api/src/main/java/org/slf4j/helpers/Util.java

```diff
@@ -64,21 +111,58 @@ private Util() {
                 break;
         }
 
-        // trace[i] = Util; trace[i+1] = caller; trace[i+2] = caller's caller
-        if (i >= trace.length || i + 2 >= trace.length) {
-            throw new IllegalStateException("Failed to find org.slf4j.helpers.Util or its caller in the stack; " + "this should not happen");
+        // Advance until Util is no longer found
+        i++;
+        for (; i < trace.length; i++) {
+          if (!thisClassName.equals(trace[i].getName()))
+              break;
         }
 
-        return trace[i + 2];
-    }
+        // trace[i] = caller; trace[i+1] = caller's caller
+        if (i + 1 >= trace.length) {
+            throw new IllegalStateException(
+                "Failed to find org.slf4j.helpers.Util or its caller in the stack; " +
+                "this should not happen");
+        }
 
-    static final public void report(String msg, Throwable t) {
-        System.err.println(msg);
-        System.err.println("Reported exception:");
-        t.printStackTrace();
+        return trace[i + 1].getName();
     }
 
-    static final public void report(String msg) {
-        System.err.println("SLF4J: " + msg);
+    /**
+     * Returns the name of the class which called the invoking method using
+     * a new exception.
+     *
+     * This is slower than {@link #getCallingClassNameFromSecurityManager()} but
+     * works even under environments where security manager creation is
+     * forbidden by policy.
+     *
+     * @return the name of the class which called the invoking method.
+     */
+    static String getCallingClassNameFromException() {
+        StackTraceElement[] trace = new Throwable().getStackTrace();
+        String thisClassName = Util.class.getName();
+
+        // Advance until Util is found
+        int i;
+        for (i = 0; i < trace.length; i++) {
+            if (thisClassName.equals(trace[i].getClassName()))
+                break;
+        }
+
+        // Advance until Util is no longer found
+        i++;
```

Comment:
After refactoring the caller finders into separate classes, this block of code is no longer needed and has been deleted. Please take a look and let me know if any changes are still required.